File: camera/frelon/test/testfreloncontrol.cpp
#include "FrelonInterface.h"
#include "CtControl.h"
#include "CtAcquisition.h"
#include "CtImage.h"
#include "CtSaving.h"
#include "AcqState.h"

using namespace lima;
using namespace std;

DEB_GLOBAL(DebModTest);

class ImageStatusCallback : public CtControl::ImageStatusCallback
{
	DEB_CLASS(DebModTest, "ImageStatusCallback");

public:
	ImageStatusCallback(CtControl& ct, AcqState& acq_state);
	virtual ~ImageStatusCallback();

protected:
	virtual void imageStatusChanged(
				const CtControl::ImageStatus& img_status);

private:
	CtControl& m_ct;
	AcqState& m_acq_state;

	int m_nb_frames;
};

ImageStatusCallback::ImageStatusCallback(CtControl& ct, AcqState& acq_state)
	: m_ct(ct), m_acq_state(acq_state), m_nb_frames(0)
{
	DEB_CONSTRUCTOR();
}

ImageStatusCallback::~ImageStatusCallback()
{
	DEB_DESTRUCTOR();
}

void ImageStatusCallback::imageStatusChanged(
				const CtControl::ImageStatus& img_status)
{
	DEB_MEMBER_FUNCT();
	DEB_PARAM() << DEB_VAR1(img_status);

	int last_acq_frame_nb = img_status.LastImageAcquired;
	int last_saved_frame_nb = img_status.LastImageSaved;

	if (last_acq_frame_nb == 0) {
		CtAcquisition *ct_acq = m_ct.acquisition();
		ct_acq->getAcqNbFrames(m_nb_frames);
	}

	if ((last_acq_frame_nb == m_nb_frames - 1) &&
	    (m_acq_state.get() == AcqState::Acquiring)) {
		DEB_ALWAYS() << "All frames acquired!";
		m_acq_state.set(AcqState::Saving);
	}

	if (last_saved_frame_nb == m_nb_frames - 1) {
		DEB_ALWAYS() << "All frames saved!";
		m_acq_state.set(AcqState::Finished);
	}
}

class FrelonAcq
{
	DEB_CLASS(DebModTest, "FrelonAcq");
public:
	FrelonAcq(int espia_dev_nb);
	~FrelonAcq();

	void initSaving(string dir, string prefix, string suffix, int idx, 
			CtSaving::FileFormat fmt, CtSaving::SavingMode mode,
			int frames_per_file);

	void setExpTime(double exp_time);
	void setNbAcqFrames(int nb_acq_frames);

	void start();
	void wait();
	void run();

	void setBin(Bin& bin);
	void setRoi(Roi& roi);

private:
	void printDefaults();

	Espia::Dev		m_edev;
	Espia::Acq		m_acq;
	Espia::BufferMgr	m_buffer_cb_mgr;
	Espia::SerialLine	m_eserline;
	Frelon::Camera		m_cam;
	BufferCtrlMgr		m_buffer_mgr;
	Frelon::Interface	m_hw_inter;
	AcqState		m_acq_state;

	CtControl		*m_ct;
	CtAcquisition		*m_ct_acq;
	CtSaving		*m_ct_saving;
	CtImage			*m_ct_image;
	CtBuffer		*m_ct_buffer;

	ImageStatusCallback	*m_img_status_cb;
};

FrelonAcq::FrelonAcq(int espia_dev_nb)
	: m_edev(espia_dev_nb), 
	  m_acq(m_edev), 
	  m_buffer_cb_mgr(m_acq), 
	  m_eserline(m_edev),
	  m_cam(m_eserline), 
	  m_buffer_mgr(m_buffer_cb_mgr),
	  m_hw_inter(m_acq, m_buffer_mgr, m_cam),
	  m_ct(NULL), m_img_status_cb(NULL)
{
	DEB_CONSTRUCTOR();

	AutoPtr<CtControl> ct = new CtControl(&m_hw_inter);

	m_ct_acq    = ct->acquisition();
	m_ct_saving = ct->saving();
	m_ct_image  = ct->image();
	m_ct_buffer = ct->buffer();

	printDefaults();

	AutoPtr<ImageStatusCallback> img_status_cb;
	img_status_cb = new ImageStatusCallback(*ct, m_acq_state);
	ct->registerImageStatusCallback(*img_status_cb);

	DEB_TRACE() << "All is OK!";
	m_ct = ct.forget();
	m_img_status_cb = img_status_cb.forget();
}

FrelonAcq::~FrelonAcq()
{
	DEB_DESTRUCTOR();

	delete m_img_status_cb;
	delete m_ct;
}

void FrelonAcq::printDefaults()
{
	DEB_MEMBER_FUNCT();

	AcqMode acq_mode;
	m_ct_acq->getAcqMode(acq_mode);
	DEB_TRACE() << "Default " << DEB_VAR1(acq_mode);

	ImageType image_type;
	m_ct_image->getImageType(image_type);
	DEB_TRACE() << "Default " << DEB_VAR1(image_type);

	Size max_size;
	m_ct_image->getMaxImageSize(max_size);
	DEB_TRACE() << "Default " << DEB_VAR1(max_size);

	CtImage::ImageOpMode img_op_mode;
	m_ct_image->getMode(img_op_mode);
	DEB_TRACE() << "Default " << DEB_VAR1(img_op_mode);

	Bin bin;
	m_ct_image->getBin(bin);
	DEB_TRACE() << "Default " << DEB_VAR1(bin);

	Roi roi;
	m_ct_image->getRoi(roi);
	DEB_TRACE() << "Default " << DEB_VAR1(roi);
	
	double exp_time;
	m_ct_acq->getAcqExpoTime(exp_time);
	DEB_TRACE() << "Default " << DEB_VAR1(exp_time);

	int nb_frames;
	m_ct_acq->getAcqNbFrames(nb_frames);
	DEB_TRACE() << "Default " << DEB_VAR1(nb_frames);

}

void FrelonAcq::start()
{
	DEB_MEMBER_FUNCT();

	m_ct->prepareAcq();
	m_acq_state.set(AcqState::Acquiring);
	DEB_TRACE() << "Starting acquisition";
   	m_ct->startAcq();
}

void FrelonAcq::wait()
{
	DEB_MEMBER_FUNCT();
	m_acq_state.waitNot(AcqState::Acquiring | AcqState::Saving);
	DEB_TRACE() << "Acquisition finished";
}

void FrelonAcq::run()
{
	DEB_MEMBER_FUNCT();
	start();
	wait();
}

void FrelonAcq::initSaving(string dir, string prefix, string suffix, int idx, 
			   CtSaving::FileFormat fmt, CtSaving::SavingMode mode,
			   int frames_per_file)
{
	DEB_MEMBER_FUNCT();

	m_ct_saving->setDirectory(dir);
 	m_ct_saving->setPrefix(prefix);
	m_ct_saving->setSuffix(suffix);
	m_ct_saving->setNextNumber(idx);
	m_ct_saving->setFormat(fmt);
	m_ct_saving->setSavingMode(mode);
	m_ct_saving->setFramesPerFile(frames_per_file);

}

void FrelonAcq::setExpTime(double exp_time)
{
	DEB_MEMBER_FUNCT();
	DEB_PARAM() << DEB_VAR1(exp_time);

	m_ct_acq->setAcqExpoTime(exp_time);
}

void FrelonAcq::setNbAcqFrames(int nb_acq_frames)
{
	DEB_MEMBER_FUNCT();
	DEB_PARAM() << DEB_VAR1(nb_acq_frames);

	m_ct_acq->setAcqNbFrames(nb_acq_frames);
}

void FrelonAcq::setBin(Bin& bin)
{
	DEB_MEMBER_FUNCT();
	DEB_PARAM() << DEB_VAR1(bin);

	m_ct_image->setBin(bin);
}

void FrelonAcq::setRoi(Roi& roi)
{
	DEB_MEMBER_FUNCT();
	DEB_PARAM() << DEB_VAR1(roi);

	m_ct_image->setRoi(roi);

}

void test_frelon_control(bool enable_debug = false)
{
	DEB_GLOBAL_FUNCT();

	if (!enable_debug) {
		DebParams::disableModuleFlags(DebParams::AllFlags);
	}

	DEB_ALWAYS() << "Creating FrelonAcq";
	int espia_dev_nb = 0;
	FrelonAcq acq(espia_dev_nb);
	DEB_ALWAYS() << "Done!";

	acq.initSaving("data", "img", ".edf", 0, CtSaving::EDF, 
		       CtSaving::AutoFrame, 1);

	DEB_ALWAYS() << "First run with default pars";
	acq.run();
	DEB_ALWAYS() << "Done!";

	double exp_time = 1e-6;
	acq.setExpTime(exp_time);

	int nb_acq_frames = 500;
	acq.setNbAcqFrames(nb_acq_frames);

	DEB_ALWAYS() << "Run " << DEB_VAR2(exp_time, nb_acq_frames);
	acq.run();
	DEB_ALWAYS() << "Done!";

	Bin bin(2);
	acq.setBin(bin);

	nb_acq_frames = 5;
	acq.setNbAcqFrames(nb_acq_frames);

	DEB_ALWAYS() << "Run " << DEB_VAR2(bin, nb_acq_frames);
	acq.run();
	DEB_ALWAYS() << "Done!";

	Roi roi = Roi(Point(256, 256), Size(512, 512));
	acq.setRoi(roi);

	DEB_ALWAYS() << "Run " << DEB_VAR1(roi);
	acq.run();
	DEB_ALWAYS() << "Done!";

	roi = Roi(Point(267, 267), Size(501, 501));
	acq.setRoi(roi);

	DEB_ALWAYS() << "Run " << DEB_VAR1(roi);
	acq.run();
	DEB_ALWAYS() << "Done!";

}

int main(int argc, char *argv[])
{
	DEB_GLOBAL_FUNCT();

	string par;
	if (argc > 1)
		par = argv[1];
	bool enable_debug = (par == "debug");

	try {
		test_frelon_control(enable_debug);
	} catch (Exception& e) {
		DEB_ERROR() << "LIMA Exception: " << e;
	}
	return 0;
}

File: camera/frelon/test/testfreloncontrol.py
import os, sys, string, gc, time
from lima import *
from Debug import *
import processlib

DEB_GLOBAL(DebModTest)

class ImageStatusCallback(CtControl.ImageStatusCallback):

    DEB_CLASS(DebModTest, "ImageStatusCallback")

    @DEB_MEMBER_FUNCT
    def __init__(self, ct, acq_state, print_time=1):
        CtControl.ImageStatusCallback.__init__(self)

        self.m_ct = ct
        self.m_acq_state = acq_state
        self.m_nb_frames = 0
        
        self.m_last_print_ts = 0
        self.m_print_time = print_time
        
    @DEB_MEMBER_FUNCT
    def imageStatusChanged(self, img_status):
        last_acq_frame_nb = img_status.LastImageAcquired;
        last_saved_frame_nb = img_status.LastImageSaved;

        if last_acq_frame_nb == 0:
            ct_acq = self.m_ct.acquisition()
            self.m_nb_frames = ct_acq.getAcqNbFrames()

        acq_state_changed = False
        msg = ''
        if ((last_acq_frame_nb == self.m_nb_frames - 1) and 
            (self.m_acq_state.get() == AcqState.Acquiring)):
            msg = 'All frames acquired!'
            self.m_acq_state.set(AcqState.Saving)
            acq_state_changed = True

        if last_saved_frame_nb == self.m_nb_frames - 1:
            msg = 'All frames saved!'
            self.m_acq_state.set(AcqState.Finished)
            acq_state_changed = True
            
        now = time.time()
        if ((now - self.m_last_print_ts >= self.m_print_time) or
            acq_state_changed):
            deb.Always("Last Acquired: %8d, Last Saved: %8d" % 
                       (last_acq_frame_nb, last_saved_frame_nb))
                      
            self.m_last_print_ts = now

        if msg:
            deb.Always(msg)


class FrelonAcq:

    DEB_CLASS(DebModTest, "FrelonAcq")
    
    @DEB_MEMBER_FUNCT
    def __init__(self, espia_dev_nb, use_events=False, print_time=1):
        self.m_edev          = Espia.Dev(espia_dev_nb)
        self.m_acq           = Espia.Acq(self.m_edev)
        self.m_buffer_cb_mgr = Espia.BufferMgr(self.m_acq)
        self.m_eserline      = Espia.SerialLine(self.m_edev)
        self.m_cam           = Frelon.Camera(self.m_eserline)
        self.m_buffer_mgr    = BufferCtrlMgr(self.m_buffer_cb_mgr)
        self.m_hw_inter      = Frelon.Interface(self.m_acq,
                                                     self.m_buffer_mgr,
                                                     self.m_cam)
        self.m_acq_state     = AcqState()
        self.m_ct            = CtControl(self.m_hw_inter)
        self.m_ct_acq        = self.m_ct.acquisition()
        self.m_ct_saving     = self.m_ct.saving()
        self.m_ct_image      = self.m_ct.image()
        self.m_ct_buffer     = self.m_ct.buffer()
        self.m_ct_display    = self.m_ct.display()

        self.m_use_events    = use_events
        self.m_print_time    = print_time
        
        if self.m_use_events:
            cb = ImageStatusCallback(self.m_ct, self.m_acq_state, print_time)
            self.m_img_status_cb = cb
            self.m_ct.registerImageStatusCallback(self.m_img_status_cb)
        else:
            self.m_poll_time = 0.1

        self.m_ct_display.setNames('_ccd_ds_', 'frelon_live')
        self.m_ct_display.setActive(True)
        

    @DEB_MEMBER_FUNCT
    def __del__(self):
        if self.m_use_events:
            del self.m_img_status_cb;	gc.collect()
            
        del self.m_ct_buffer, self.m_ct_image, self.m_ct_saving, self.m_ct_acq
        del self.m_ct;			gc.collect()
        del self.m_acq_state;		gc.collect()
        del self.m_hw_inter;		gc.collect()
        del self.m_buffer_mgr;		gc.collect()
        del self.m_cam;			gc.collect()
        del self.m_eserline;		gc.collect()
        del self.m_buffer_cb_mgr;	gc.collect()
        del self.m_acq;			gc.collect()
        del self.m_edev;		gc.collect()

    @DEB_MEMBER_FUNCT
    def start(self):
        self.m_ct.prepareAcq()
        self.m_acq_state.set(AcqState.Acquiring)
        self.m_ct.startAcq()

    @DEB_MEMBER_FUNCT
    def wait(self):
        if self.m_use_events:
            state_mask = AcqState.Acquiring | AcqState.Saving
            self.m_acq_state.waitNot(state_mask)
        else:
            nb_frames = self.m_ct_acq.getAcqNbFrames()
            last_print_ts = 0
            running_states = [AcqState.Acquiring, AcqState.Saving]
            while self.m_acq_state.get() in running_states:
                img_status = self.m_ct.getImageStatus()
                last_acq_frame_nb = img_status.LastImageAcquired;
                last_saved_frame_nb = img_status.LastImageSaved;

                acq_state_changed = False
                msg = ''
                if ((last_acq_frame_nb == nb_frames - 1) and 
                    (self.m_acq_state.get() == AcqState.Acquiring)):
                    msg = 'All frames acquired!'
                    self.m_acq_state.set(AcqState.Saving)
                    acq_state_changed = True

                if last_saved_frame_nb == nb_frames - 1:
                    msg = 'All frames saved!'
                    self.m_acq_state.set(AcqState.Finished)
                    acq_state_changed = True
            
                now = time.time()
                if ((now - last_print_ts >= self.m_print_time) or
                    acq_state_changed):
                    deb.Always("Last Acquired: %8d, Last Saved: %8d" % 
                               (last_acq_frame_nb, last_saved_frame_nb))
                    last_print_ts = now

                if msg:
                    print msg

                time.sleep(self.m_poll_time)

        pool_thread_mgr = processlib.PoolThreadMgr.get()
        pool_thread_mgr.wait()

    @DEB_MEMBER_FUNCT
    def run(self):
        self.start()
        self.wait()

    @DEB_MEMBER_FUNCT
    def initSaving(self, dir, prefix, suffix, idx, fmt, mode, frames_per_file):
        self.m_ct_saving.setDirectory(dir)
        self.m_ct_saving.setPrefix(prefix)
        self.m_ct_saving.setSuffix(suffix)
        self.m_ct_saving.setNextNumber(idx)
        self.m_ct_saving.setFormat(fmt)
        self.m_ct_saving.setSavingMode(mode)
        self.m_ct_saving.setFramesPerFile(frames_per_file)
        
    @DEB_MEMBER_FUNCT
    def setExpTime(self, exp_time):
        self.m_ct_acq.setAcqExpoTime(exp_time)

    @DEB_MEMBER_FUNCT
    def setNbAcqFrames(self, nb_acq_frames):
        self.m_ct_acq.setAcqNbFrames(nb_acq_frames)

    @DEB_MEMBER_FUNCT
    def setBin(self, bin):
        self.m_ct_image.setBin(bin)

    @DEB_MEMBER_FUNCT
    def setRoi(self, roi):
        self.m_ct_image.setRoi(roi)


@DEB_GLOBAL_FUNCT
def test_frelon_control(enable_debug):

    if not enable_debug:
        DebParams.disableModuleFlags(DebParams.AllFlags)

    deb.Always("Creating FrelonAcq")
    espia_dev_nb = 0
    use_events = False
    acq = FrelonAcq(espia_dev_nb, use_events)
    deb.Always("Done!")
    
    acq.initSaving("data", "img", ".edf", 0, CtSaving.EDF, 
                   CtSaving.AutoFrame, 1);

    deb.Always("First run with default pars")
    acq.run()
    deb.Always("Done!")
    
    exp_time = 1e-6
    acq.setExpTime(exp_time)

    nb_acq_frames = 500
    acq.setNbAcqFrames(nb_acq_frames)

    deb.Always("Run exp_time=%s, nb_acq_frames=%s" % (exp_time, nb_acq_frames))
    acq.run()
    deb.Always("Done!")
    
    bin = Bin(2, 2)
    acq.setBin(bin)

    nb_acq_frames = 5
    acq.setNbAcqFrames(nb_acq_frames)

    deb.Always("Run bin=<%sx%s>, nb_acq_frames=%s" % 
               (bin.getX(), bin.getY(), nb_acq_frames))
    acq.run()
    deb.Always("Done!")
    
    roi = Roi(Point(256, 256), Size(512, 512));
    acq.setRoi(roi);

    roi_tl, roi_size = roi.getTopLeft(), roi.getSize()
    deb.Always("Run roi=<%s,%s>-<%sx%s>" %
               (roi_tl.x, roi_tl.y,
                roi_size.getWidth(), roi_size.getHeight()))
    acq.run()
    deb.Always("Done!")
    
    roi = Roi(Point(267, 267), Size(501, 501));
    acq.setRoi(roi);

    roi_tl, roi_size = roi.getTopLeft(), roi.getSize()
    deb.Always("Run roi=<%s,%s>-<%sx%s>" %
               (roi_tl.x, roi_tl.y,
                roi_size.getWidth(), roi_size.getHeight()))
    acq.run()
    deb.Always("Done!")
    

def main(argv):

    enable_debug = False
    if len(argv) > 1:
        enable_debug = (argv[1] == 'debug')

    test_frelon_control(enable_debug)
        
        
    

if __name__ == '__main__':
    main(sys.argv)
